docs/proposals/20200506-conditions.md

```diff
@@ -0,0 +1,573 @@
+```
+
+In other cases, the combination of `Reason` and `Severity` allows to detect when a failure is due to a catastrophic
+error or to other events that are transient or can be eventually remediated by an user intervention
+
+```
+MachineReady=False, Reason=MachineNotHealthy, Severity=Error
+MachineReady=False, Reason=MachineUnderProbation, Severity=Warning
+AWSMachineReady=False, Reason=SSHKeyMissing, Severity=Warning
+```
+
+A more contextualized example/explanation about the Severity field can be found in the
+“Cluster provisioning” example later on in this document.
+
+#### Constraints
+In order to ensure a consistent implementation of conditions across all the Cluster API components
+the following constraints/design principles MUST be applied:
+
+##### Condition semantic
+
+- Condition types in Cluster API core objects MUST be provider agnostic. 
+- Condition types in Cluster API core objects MUST represent the operational state of a component
+  in the cluster, where the operational state is when the component is ready to serve application workloads. It should be avoided to represent only the infrastructure/provision part of the component's lifecycle.
+- Condition types in Cluster API provider-specific objects COULD be used for surfacing more
+  granular/internal details about provisioning, but always with a user driven perspective (conditions != debug).
+- Operations like upgrades, scaling-up, or scaling-down, even if orchestrated in a non-disruptive fashion,
+  MUST be considered as a deviation from the normal operational state of the cluster, and the operator
+  should be always informed when those changes are happening. 
+  e.g. There should be a condition with Severity=Warning.
+
+##### The Ready condition
+
+- A `Ready` condition SHOULD be provided at object level to represent _the overall_ operational state of
+  the component (NB. a more contextualized example/explanation of the Ready conditions can be found
+  in the “Cluster provisioning” example later on in this document).
+- The `Ready` condition MUST be based on the summary of more detailed conditions existing on the same
+  object, if defined. 
+  e.g. `AWSCluster.Status.Conditions[Ready]` condition should be the summary of `VPCReady`, `SubnetsReady`,
+  `InternetGatewaysReady` conditions.
+- Detailed conditions at object level COULD be based on `Ready` conditions gathered from the dependent
+  objects. e.g. 
+  - `Cluster.Status.Conditions[ClusterInfrastructureReady]` is a condition based on infrastructure
+  cluster state, e.g. `AWSCluster.Status.Conditions[Ready]`. 
+  - `KubeadmControlPlane.Status.Conditions[MachinesReady]` is a condition based on the aggregation
+  of the underlying `Machines.Status.Conditions[Ready]` conditions.
+- A corollary of the above set of constraints is that an object SHOULD NEVER be in status `Ready=True`
+  if one of the object's conditions are `false` or if one of the object dependents is in status `Ready=False`.
```

Comment:
Does this mean that `Cluster.Status.Conditions[Ready]` will be False if any Machine belonging to the Cluster has `Machine.Status.Conditions[Ready]` = False? 

If so, I'm not sure that I fully agree there, since the Cluster can be fully functional even if some subset of the components that make up the Cluster are not (and for any sufficiently large cluster a degraded state of some kind would be more normal than it would be exceptional).

Reply:
@detiber, I agree with your concern about the need of finding a good rule for what we report all the machines at cluster level, and this is why later in the document, when we discuss WorkerReady, we are stating that "_Considering the complexity and to wait for more in-depth analysis and use-cases, the implementation of this Condition is deferred until a later stage_."

With that in mind, we are embedding following flexibility point in the document:
- each object SHOULD (instead if MUST) have a ready condition (so e.g. during implementation we can choose to not create an overall Ready condition at cluster level)
- the Ready condition state SHOULD (instead if MUST) be consistent with the state of underlying source conditions (so in future there is room to qualify "a degraded state of some kind", like e.g. Info if failures < x%, Warning if ... etc.)